[Team-25 데이먼 &  벅픽] Side-dish 1주차 수요일 PR (프로젝트 초기설정, DB 설계, API 설계, Entity 클래스 생성)

bugpigg opened the pull request:
안녕하세요~~
 25조 데이먼 & 벅픽입니다!
현재까지 DB 설계, API 설계, 프로젝트 초기설정, Entity 클래스 생성에 대한 PR  입니다. 
초기에 DB 및 API 설계에 집중하여, 작업량이 많지 않습니다.
2주동안 잘 부탁드립니다!

<details>
<summary><h2>DB설계</h2></summary>
<div markdown="1">

초기 ERD 및 DB 테이블입니다.

![image](https://user-images.githubusercontent.com/91416897/163978964-6f4ae284-7b71-4c6c-a09a-aa2027637688.png)

![sidedish (3)](https://user-images.githubusercontent.com/91416897/163979135-0a94ace5-2e1e-49f2-b450-e30de5b01e72.png)

초기 클래스다이어그램을 그려보면서, 비즈니스 로직을 설계하였습니다.

https://cookie-giant-a00.notion.site/BE-Class-Diagram-with-Business-logic-0060f51a1c1c480e8b5de0959c248c19

</div>
</details>

<details>
<summary><h2>API 설계</h2> </summary>
<div markdown="1">

API 초기 설계입니다.
[API 설계 · dukcode/sidedish Wiki (github.com)](https://github.com/dukcode/sidedish/wiki/API-%EC%84%A4%EA%B3%84)

</div>
</details>

### 금요일까지의 계획
  - mock API 서버 제공
  - Repository 레이어 구현 및 테스트 코드 작성
  - Service, Controller 레이어 구현(기본 API 동작)

dukcode committed:
build: 스프링 부트 초기 설정 (#4)

- 의존성 Spring WEB, Validation, Lombok, Spring Data JDBC 추가
- gitignore 설정

bugpigg committed:
build: 스프링 부트 추가 프로젝트 설정 (#51)

* (#43)build: application.yml 설정

  - 로컬용 프로필을 설정하였습니다.

* (#44)build: logback-spring.xml 설정

  - 로컬용 logging 프로필을 설정하였습니다.

bugpigg committed:
feat: 전역 예외 처리 기능 추가 (#53)

* (#46)feat: 반환할 예외 DTO 클래스인 ErrorResponse 추가

* (#47)feat: 전역 예외 처리 클래스인 GlobalExceptionHandler 추가

* (#48)feat: 전역 예외 클래스인 GlobalRuntimeException 추가

dukcode committed:
feat: Entity 구현 (#58)

* (#6)feat: Product Entity 구현

* (#7)feat: Category Entity 구현

* (#8)feat: Event Entity 구현

* (#10)feat: Delivery Entity 구현

- DeliveryType Enum 클래스 추가

* (#9)feat: ProductDelivery Entity 구현

* (#54)feat: ProductEvent Entity 구현

* (#55)feat: Image Entity 구현

dukcode committed:
refactor: Entity Getter 삭제

dukcode committed:
build: application.yml h2 url 변경

dukcode committed:
build: DDL, DML 초기 데이터 설정 (#62)

* (#60)build: 테이블 DDL 생성

* (#61)build: 초기화 데이터 DML.sql 추가

- application.yml 에 경로 추가

wheejuni commented on BE/HELP.md:
```diff
@@ -0,0 +1,26 @@
+# Getting Started
```
오 이런 파일은 빼도 되지 않겠습니까? 어떻게 생각하세요? ㅎㅎ

bugpigg replied:
넵 맞습니다..ㅎㅎ 수정하겠습니다!!

wheejuni commented on BE/src/main/java/com/team25/sidedish/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,13 @@
+@AllArgsConstructor
+public class ErrorResponse {
+
+    private final int statusCode;
```
코드가 헤더로 내려갈텐데 꼭 필드로써 필요한지는.. 의문입니다~

bugpigg replied:
생각해보니 응답하는 JSON에서  `statusCode` 는 불필요한 정보였네요! 수정하겠습니다!

wheejuni commented on BE/src/main/java/com/team25/sidedish/exception/GlobalRuntimeException.java:
```diff
@@ -0,0 +1,12 @@
+
+import org.springframework.http.HttpStatus;
+
+public abstract class GlobalRuntimeException extends RuntimeException {
```
`Global` 이라는 접두사는 어색합니다~

bugpigg replied:
이 클래스를 생성한 이유는 다음과 같았습니다.

추후 각 도메인에서 발생가능한 예외들을 사용자 정의 예외로 정의할때, 이 `GlobalRuntimeException` 클래스를 상속받게 하여 `GlobalExceptionHandler` 에서 일괄적으로 처리해주기 위함이였습니다. 

그래서 `Global`접두사를 사용하게 되었는데,  코멘트를 통해 생각해보니 사용자 정의 예외를 일괄적으로 처리하는 것보다는 세분화하여 처리하는게 더 적절하다고 생각되네요! 이 부분도 수정하겠습니다!

wheejuni commented on BE/src/main/java/com/team25/sidedish/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
`global` 이라는 표현을 좀 남용하시는 느낌이 있는 것 같은데 이렇게 작명을 하면 클래스가 가진 책임이나 역할 범위가 굉장히 모호해집니다. 이런 네이밍 없었으면 하고요.

지금 초기 셋업 단계라서 이렇게 하신 거겠지만 예외는 최대한 세분화한 핸들링 로직이 있었으면 합니다.

bugpigg replied:
`최대한 세분화한 예외 핸들링 로직`  을 구현하려 노력해보겠습니다!

wheejuni commented on BE/src/main/java/com/team25/sidedish/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,18 @@
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(value = GlobalRuntimeException.class)
```
너무 광범위합니다. 크게 의미가 없는 듯 보여요.

wheejuni commented on BE/src/main/java/com/team25/sidedish/exception/GlobalRuntimeException.java:
```diff
@@ -0,0 +1,12 @@
+        super(message);
+    }
+
+    public abstract HttpStatus getStatus();
```
예외 클래스에 `org.springframework..` 로 시작하는 의존성 있으면 안된다고 생각합니다.
웹 MVC 프레임워크는 언제든지 교체할 수 있어야 합니다~ 나중에 다른거 쓰면, 여기까지 손대야 할까요?

bugpigg replied:
Spring framework와의 의존성까지는 미처 생각하지 못했었네요.
이 부분도 의존성을 제거할 방법을 생각해보겠습니다!

wheejuni reviewed (changes requested):
대단히 수고 많으셨습니다 💯 
코멘트 몇 개 남겼으니 확인 부탁드립니다

bugpigg committed:
refactor: 1차 PR 코멘트 반영 (#63)

* docs: HELP.md 제거

* fix: ErrorResponse DTO 에서 불필요한 필드 제거

- ErrorResponse DTO 에서 불필요한 필드였던 statusCode 를 제거하였습니다.

* fix: GlobalRuntimeException 제거

- 예외들을 세분화하여 관리하기위해 GlobalRuntimeException 을 제거하였습니다.

* fix: GlobalExceptionHandler 제거

- 추후 세분화 하여 ExceptionHandler 구현하기 위해, GlobalExceptionHandler 를 제거하였습니다.

* feat: NotFoundException 생성

wheejuni reviewed (approved):
👍 다음단계로 넘어갑시다
